src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -151,88 +154,108 @@ class GroupArrayNumericImpl final
         }
 
         if constexpr (Trait::sampler == Sampler::RNG)
         {
-            auto & a = this->data(place);
-            ++a.total_values;
-            if (a.value.size() < max_elems)
-                a.value.push_back(assert_cast<const ColumnVector<T> &>(*columns[0]).getData()[row_num], arena);
+            if (cur_elems.value.size() < max_elems)
+                cur_elems.value.push_back(row_value, arena);
             else
             {
-                UInt64 rnd = a.genRandom(a.total_values);
+                UInt64 rnd = cur_elems.genRandom(cur_elems.total_values);
                 if (rnd < max_elems)
-                    a.value[rnd] = assert_cast<const ColumnVector<T> &>(*columns[0]).getData()[row_num];
+                    cur_elems.value[rnd] = row_value;
             }
         }
-        // TODO
-        // if constexpr (Trait::sampler == Sampler::DETERMINATOR)
     }
 
     void merge(AggregateDataPtr __restrict place, ConstAggregateDataPtr rhs, Arena * arena) const override
     {
-        if constexpr (Trait::sampler == Sampler::NONE)
-        {
-            auto & cur_elems = this->data(place);
-            auto & rhs_elems = this->data(rhs);
+        auto & cur_elems = this->data(place);
+        auto & rhs_elems = this->data(rhs);
 
-            if (!limit_num_elems)
-            {
-                if (rhs_elems.value.size())
-                    cur_elems.value.insertByOffsets(rhs_elems.value, 0, rhs_elems.value.size(), arena);
-            }
-            else
-            {
-                UInt64 elems_to_insert = std::min(static_cast<size_t>(max_elems) - cur_elems.value.size(), rhs_elems.value.size());
-                if (elems_to_insert)
-                    cur_elems.value.insertByOffsets(rhs_elems.value, 0, elems_to_insert, arena);
-            }
-        }
+        if (rhs_elems.value.empty())
+            return;
 
-        if constexpr (Trait::sampler == Sampler::RNG)
+        if constexpr (Trait::last)
+            mergeNoSamplerLast(cur_elems, rhs_elems, arena);
+        else if constexpr (Trait::sampler == Sampler::NONE)
+            mergeNoSampler(cur_elems, rhs_elems, arena);
+        else if constexpr (Trait::sampler == Sampler::RNG)
+            mergeWithRNGSampler(cur_elems, rhs_elems, arena);
+    }
+
+    void mergeNoSamplerLast(Data & cur_elems, const Data & rhs_elems, Arena * arena) const
+    {
+        UInt64 new_elements = std::min(max_elems, cur_elems.value.size() + rhs_elems.value.size());
+        cur_elems.value.resize_exact(new_elements, arena);
+        for (auto & value : rhs_elems.value)
         {
-            if (this->data(rhs).value.empty()) /// rhs state is empty
-                return;
+            cur_elems.value[cur_elems.total_values % max_elems] = value;
+            ++cur_elems.total_values;
+        }
+        assert(rhs_elems.total_values >= rhs_elems.value.size());
+        cur_elems.total_values += rhs_elems.total_values - rhs_elems.value.size();
```

Comment:
`total_values` is used to determine the next position for "circular buffer" for `groupArrayLast`.
During merge, in the loop above, it will take into account only `max_elems` values from the `rhs`, and the total position should be adjusted to take into account position in "circular buffer" of the `rhs`.